src/core/Akka/Actor/ActorCell.cs

```diff
@@ -308,33 +308,31 @@ public virtual void SendMessage(IActorRef sender, object message)
             if (_systemImpl.Settings.SerializeAllMessages)
             {
                 DeadLetter deadLetter;
-                var unwrapped = (deadLetter = message as DeadLetter) != null ? deadLetter.Message : message;
+                var unwrapped = (deadLetter = message.Message as DeadLetter) != null ? deadLetter.Message : message.Message;
                 if (!(unwrapped is INoSerializationVerificationNeeded))
                 {
-                    Serializer serializer = _systemImpl.Serialization.FindSerializerFor(message);
-                    byte[] serialized = serializer.ToBinary(message);
-
+                    Serializer serializer = _systemImpl.Serialization.FindSerializerFor(message.Message);
+                    byte[] serialized = serializer.ToBinary(message.Message);
 
                     var manifestSerializer = serializer as SerializerWithStringManifest;
                     if (manifestSerializer != null)
                     {
                         var manifest = manifestSerializer.Manifest(serialized);
-                        message = _systemImpl.Serialization.Deserialize(serialized, manifestSerializer.Identifier, manifest);
+                        message = new Envelope(_systemImpl.Serialization.Deserialize(serialized, manifestSerializer.Identifier, manifest), message.Sender);
```

Comment:
Scala version has a bit different implementation
https://github.com/akka/akka/blob/4acc1cca6a27be0ff80f801de3640f91343dce94/akka-actor/src/main/scala/akka/actor/dungeon/Dispatch.scala#L138

I should check for null, and then throw an exception